clang/test/SemaSYCL/unique-stable-id.cpp

```diff
@@ -0,0 +1,123 @@
+// RUN: %clang_cc1 %s -std=c++17 -triple x86_64-pc-windows-msvc -fsycl-is-device -verify -fsyntax-only -Wno-unused
+// RUN: %clang_cc1 %s -std=c++17 -triple x86_64-linux-gnu -fsycl-is-device -verify -fsyntax-only -Wno-unused
+// Various Semantic analysis tests for the __builtin_unique_stable_id feature.
+
+#include "Inputs/sycl.hpp"
+
+struct S {}; // #SDecl
+int f();
+
+template<auto &S>
+void wrapper() {
+  __builtin_sycl_unique_stable_id(S);
+}
+
+
+static constexpr double global_double = 0;
+
+template<typename T>
+void usage_templ(T &t) {
+  // expected-error@+2{{'__builtin_sycl_unique_stable_id' variable must have global storage}}
+  // expected-note@#usage_templ_instantiation{{in instantiation of function template specialization}}
+  __builtin_sycl_unique_stable_id(t);
+}
+
+void usage(S s) {
+  usage_templ(s); // #usage_templ_instantiation
+  // expected-error@+2{{'S' does not refer to a value}}
+  // expected-note@#SDecl{{declared here}}
+  __builtin_sycl_unique_stable_id(S);
+  // expected-error@+1{{expected variable name}}
+  __builtin_sycl_unique_stable_id(f);
+
+  // expected-error@+1{{expected variable name}}
+  __builtin_sycl_unique_stable_id(f());
+
+  // Needs to work.
+  wrapper<global_double>();
```

Comment:
Huh, do `static constexpr` variables actually *have* global storage? I guess I always assumed those were compile-time constants with no runtime storage, but I've never really looked that closely at them.

Reply:
When exported, they have global storage, since you can do things like take an address of them.  The difference is they are basically 'inline' in those cases.

Reply:
Ahhh, thanks for the info!